c7n/resources/iam.py

```diff
@@ -795,19 +798,29 @@ def process(self, resources):
         client = local_session(self.manager.session_factory).client('iam')
         policy_arn = self.data['arn']
         state = self.data['state']
-
+        if state == 'attached' and policy_arn == "*":
+            raise PolicyValidationError(
+                'State should be detached when arn is * on %s' % (self.manager.data))
         for r in resources:
             if state == 'attached':
                 client.attach_role_policy(
                     RoleName=r['RoleName'],
                     PolicyArn=policy_arn)
-            elif state == 'detached':
+            elif state == 'detached' and not policy_arn == "*":
                 try:
                     client.detach_role_policy(
                         RoleName=r['RoleName'],
                         PolicyArn=policy_arn)
                 except client.exceptions.NoSuchEntityException:
                     pass
+            elif state == 'detached' and policy_arn == "*":
+                attached_policy = client.list_attached_role_policies(RoleName=r['RoleName'])
+                policy_arns = [p.get('PolicyArn') for p in attached_policy['AttachedPolicies']]
+                for parn in policy_arns:
+                    try:
+                        client.detach_role_policy(RoleName=r['RoleName'], PolicyArn=parn)
+                    except client.exceptions.NoSuchEntityException:
```

Comment:
so rather than wrapping each individual api call with the exception handler this loop body for processing a role resource should turn into a method and then we can just try/except around invoking that method.

Reply:
Can do that. But in that case afaics, there'll be two sets of try/except blocks, one in the method and the other where we invoke the method. Is that fine?

Reply:
just need one when you invoke, ie move from call sites to around the method encompassing all the call sites.

Reply:
Thanks @kapilt! Addressed.